core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java

```diff
@@ -0,0 +1,279 @@
+            for (WhenClause whenClause : caseExpression.getWhenClauses()) {
+                Expression whenOperand = whenClause.getOperand();
+                if (!(whenOperand instanceof ComparisonExpression)) {
+                    return false;
+                }
+                ComparisonExpression whenComparisonFunction = (ComparisonExpression) whenOperand;
+                Expression left = whenComparisonFunction.getLeft();
+                Expression right = whenComparisonFunction.getRight();
+
+                if (!whenComparisonFunction.getOperator().equals(EQUAL)) {
+                    return false;
+                }
+
+                if (commonLHSExpression == null) {
+                    commonLHSExpression = left;
+                }
+                else if (!commonLHSExpression.equals(left)) {
+                    return false;
+                }
+                rightHandSideExpressions.add(right);
+            }
+            return checkUniqueLiterals(rightHandSideExpressions);
+        }
+
+        private boolean checkUniqueLiterals(List<Expression> expressions)
+        {
+            Set<Expression> literals = new HashSet<>();
+            for (Expression expression : expressions) {
+                if (!isLiteral(expression) || literals.contains(expression)) {
+                    return false;
+                }
+                literals.add(expression);
+            }
+            return true;
+        }
+
+        private boolean isLiteral(Expression expression)
+        {
+            if (expression instanceof Cast) {
+                return isLiteral(((Cast) expression).getExpression());
+            }
+            return expression instanceof Literal;
+        }
+
+        private Expression processCaseExpression(
+                Optional<Cast> castExpression,
+                List<WhenClause> whenClauses,
+                Optional<Expression> defaultValue,
+                Function<WhenClause, Expression> operandExtractor,
+                Expression otherExpression,
+                ComparisonExpression.Operator operator)
+        {
+            List<Expression> andExpressions = new ArrayList<>();
+            List<Expression> invertedOperands = new ArrayList<>();
+
+            for (WhenClause whenClause : whenClauses) {
+                Expression whenOperand = operandExtractor.apply(whenClause);
+                Expression whenResult = getCastExpression(castExpression, whenClause.getResult());
+
+                andExpressions.add(ExpressionUtils.and(
+                        new ComparisonExpression(operator, whenResult, otherExpression),
+                        whenOperand));
+                invertedOperands.add(new NotExpression(whenOperand));
+            }
+
+            Expression defaultExpression = defaultValue
+                    .map(value -> getCastExpression(castExpression, value))
+                    .orElse(getCastExpression(castExpression, new NullLiteral()));
```

Comment:
For consistency with `LiteralEncoder`, cast of `NullLiteral` should be not safe and type-only. Please use the `LiteralEncoder.toExpression()` method to get the proper form of `Cast` over `NullLiteral`.

Reply:
Done.